File: GuideObjC/GuideObjC/Views/FileImageView.m
//
//  FileImageView.m
//  GuideObjC
//
//  Created by LuzanovRoman on 16.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

#import "FileImageView.h"
#import "SimpleFileManager.h"
#import "UIColor+Additions.h"

@interface FileImageView ()
@property (retain, nonatomic) UIActivityIndicatorView *indicator;
@end

@implementation FileImageView

- (id)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    
    if (self) {
        _indicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhite];
        _indicator.translatesAutoresizingMaskIntoConstraints = false;
        _indicator.color = [UIColor hexColor:0x5E66B1];
        [self addSubview:_indicator];
        [self addConstraint:[NSLayoutConstraint constraintWithItem:_indicator attribute:NSLayoutAttributeCenterX relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeCenterX multiplier:1 constant:0]];
        [self addConstraint:[NSLayoutConstraint constraintWithItem:_indicator attribute:NSLayoutAttributeCenterY relatedBy:NSLayoutRelationEqual toItem:self attribute:NSLayoutAttributeCenterY multiplier:1 constant:0]];
    }
    return self;
}

- (void)setFile:(EGFFile *)file {
    _file = file;
    
    self.image = nil;
    [self.indicator stopAnimating];
    
    if (!file) {
        return;
    }
    [self.indicator startAnimating];
    
    __block EGFFile * theFile = file;
    
    [[SimpleFileManager sharedInstance] imageWithFile:file completion:^(UIImage *image, BOOL fromCache) {
        // We must be ensure we show appropriate image
        if (theFile != _file) {
            return;
        }
        self.image = image;
        [self.indicator stopAnimating];
        
        if (image && !fromCache) {
            self.alpha = 0;
            [UIView animateWithDuration:0.3 animations:^{
                self.alpha = 1;
            } completion:^(BOOL finished) {
                self.alpha = 1;
            }];
        }
    }];
}
@end

File: GuideSwift/GuideSwift/Views/FileImageView.swift
//
//  FileImageView.swift
//  GuideSwift
//
//  Created by LuzanovRoman on 16.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

import UIKit

class FileImageView: UIImageView {
    
    fileprivate var indicator: UIActivityIndicatorView!
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        indicator = UIActivityIndicatorView(activityIndicatorStyle: .white)
        indicator.translatesAutoresizingMaskIntoConstraints = false
        indicator.color = UIColor.hexColor(0x5E66B1)
        addSubview(indicator)
        addConstraint(NSLayoutConstraint(item: indicator, attribute: .centerX, relatedBy: .equal, toItem: self, attribute: .centerX, multiplier: 1, constant: 0))
        addConstraint(NSLayoutConstraint(item: indicator, attribute: .centerY, relatedBy: .equal, toItem: self, attribute: .centerY, multiplier: 1, constant: 0))
    }
    
    var file: EGFFile? {
        didSet {
            image = nil
            indicator.stopAnimating()
            
            guard let theFile = file else { return }
            
            indicator.startAnimating()

            SimpleFileManager.shared.image(withFile: theFile) { [weak self] (image, fromCache) in
                
                guard let strongSelf = self else { return }
                
                // We must be ensure we show appropriate image
                if theFile !== strongSelf.file { return }
                
                strongSelf.image = image
                strongSelf.indicator.stopAnimating()
                
                guard let _ = image else { return }
                
                if !fromCache {
                    strongSelf.alpha = 0
                    
                    UIView.animate(withDuration: 0.3, animations: {
                        strongSelf.alpha = 1
                    }) { (finished) in
                        strongSelf.alpha = 1
                    }
                }
            }
        }
    }
}
